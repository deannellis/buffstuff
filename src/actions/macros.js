import database from '../firebase/firebase';

export const setMacros = (current = {}, target = {}) => ({
	type: 'SET_MACROS',
	target,
	current,
});

export const startFetchMacros = () => {
	return (dispatch) => {
		return database
			.ref('macros')
			.once('value')
			.then((snapshot) => {
				if (snapshot.val() !== null) {
					const { target, current } = snapshot.val();
					dispatch(setMacros(current, target));
				}
			});
	};
};

export const setTargetMacros = ({ protein, carbs, fat }) => ({
	type: 'SET_TARGET_MACROS',
	target: {
		protein,
		carbs,
		fat,
	},
});

export const startSetTargetMacros = ({
	protein = 0,
	carbs = 0,
	fat = 0,
} = {}) => {
	return (dispatch) => {
		return database
			.ref('macros')
			.set({
				target: { protein, carbs, fat },
				current: { protein: 0, carbs: 0, fat: 0 },
			})
			.then(() => {
				dispatch(setTargetMacros({ protein, carbs, fat }));
			});
	};
};

export const updateMacro = ({ macro, amount }) => ({
	type: 'UPDATE_MACRO',
	update: {
		macro,
		amount,
	},
});

export const startUpdateMacro = ({ macro = '', amount = 0 } = {}, current) => {
	return (dispatch) => {
		const newAmount = current[macro] + amount;
		return database
			.ref(`macros/current/${macro}`)
			.set(newAmount)
			.then(() => {
				dispatch(updateMacro({ macro, amount }));
			});
	};
};

export const setCurrentDate = ({ currentDate = 0 } = {}) => ({
	type: 'SET_CURRENT_DATE',
	currentDate,
});

export const resetCurrent = () => ({ type: 'RESET_CURRENT' });

export const startResetCurrent = () => {
	return (dispatch) => {
		return database
			.ref('macros/current')
			.set({ protein: 0, carbs: 0, fat: 0 })
			.then(() => {
				dispatch(resetCurrent());
			});
	};
};

// export const syncMacros = () => ({ type: 'SYNC_MACROS' });

// export const startSyncMacros = () => {
// 	return (dispatch) => {
// 		console.log('called!!!');
// 		return database
// 			.ref('macros')
// 			.set(macrosReducerDefaultState)
// 			.then(() => {
// 				dispatch(syncMacros());
// 			});
// 	};
// };
