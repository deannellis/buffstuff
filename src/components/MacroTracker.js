import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Button from './Button';
import PieChart from './PieChart';
import { DotsIcon } from '../utils/icons';
import UpdateMacroForm from '../forms/UpdateMacroForm';
import SetMacrosForm from '../forms/SetMacrosForm';

const getPieSlices = (target, current) => {
	let leftover = 0;
	const pieData = [];
	const keys = Object.keys(current);

	keys.forEach((key) => {
		const item = {};
		item.macro = key;
		if (current[key] !== null) {
			if (current[key] <= target[key]) {
				item.amount = current[key];
				const diff = target[key] - current[key];
				leftover += diff;
			} else {
				item.amount = target[key];
			}
		} else {
			item.amount = 0;
			leftover += target[key];
		}
		pieData.push(item);
	});

	if (leftover !== 0) {
		const leftoverObject = {
			macro: 'leftover',
			amount: leftover,
		};
		pieData.push(leftoverObject);
	}
	return pieData;
};

class MacroTracker extends Component {
	constructor(props) {
		super(props);
		// const { target, current } = this.props.macros;
		this.state = {
			displayUpdateForm: false,
			displayOptionsMenu: false,
			// data: getPieSlices(target, current),
		};
	}

	onSubmitMacros = (macros) => {
		const { updateMacro } = this.props;
		this.setState({ displayUpdateForm: false });
		updateMacro(macros);
	};

	onSetMacros = (macros, closeMenu) => {
		const { setMacros } = this.props;
		setMacros(macros);
		if (closeMenu) this.setState({ displayOptionsMenu: false });
	};

	render() {
		const { macros } = this.props;
		const { target, current } = macros;
		const { displayOptionsMenu, displayUpdateForm } = this.state;
		const targetKeys = Object.keys(target);

		return (
			<div className="macro-tracker card">
				<div className="macro-tracker__header">
					<h2>Macros</h2>
					<button
						onClick={() => {
							this.setState({ displayOptionsMenu: true });
						}}
						className="macro-tracker__menu-button"
						type="button"
					>
						<DotsIcon size={24} />
					</button>
				</div>
				{target.protein === null ? (
					<>
						<p>Enter Your Macros</p>
						<p className="macro-tracker__helper-text">
							Not sure what your macros are?
							<br />
							There are{' '}
							<a
								href="https://www.google.com/search?q=calculating+macros+for+muscle+gain"
								target="blank"
							>
								several calculators
							</a>{' '}
							online.
						</p>
						<SetMacrosForm
							submitMacros={(enteredMacros) => {
								const { setMacros } = this.props;
								setMacros(enteredMacros);
							}}
						/>
					</>
				) : (
					<>
						<PieChart data={getPieSlices(target, current)} />
						<div className="macro-tracker__macros">
							{targetKeys.map((macro, i) => (
								<div className="macro-tracker__macro" key={macro}>
									<div
										className="macro-tracker__macro-key"
										id={`macro-key-${i}`}
									/>
									<p>
										<span>{macro}</span>
										{` ${current[macro] !== null ? current[macro] : 0} of ${
											target[macro]
										} grams`}
									</p>
								</div>
							))}
						</div>
						<Button
							clickHandler={() => {
								this.setState({ displayUpdateForm: true });
							}}
							variant="primary"
						>
							Add macros
						</Button>
						<Button
							clickHandler={() => {
								const { resetMacros } = this.props;
								resetMacros('current');
							}}
						>
							reset macros
						</Button>
					</>
				)}
				<div
					className={
						displayUpdateForm
							? 'macro-tracker__overlay card__overlay'
							: 'macro-tracker__overlay card__overlay  card__overlay--hidden'
					}
				>
					<UpdateMacroForm
						updateMacro={this.onSubmitMacros}
						closeForm={() => {
							this.setState({ displayUpdateForm: false });
						}}
					/>
				</div>
				<div
					className={
						displayOptionsMenu
							? 'macro-tracker__overlay card__overlay macro-tracker__menu'
							: 'macro-tracker__overlay card__overlay  card__overlay--hidden macro-tracker__menu'
					}
				>
					<p>Update Target Macros</p>
					<SetMacrosForm
						submitMacros={this.onSetMacros}
						current={macros.target}
					/>
					<div className="macro-tracker__menu-bottom">
						<Button
							clickHandler={() => {
								this.setState({ displayOptionsMenu: false });
							}}
						>
							Close Menu
						</Button>
					</div>
				</div>
			</div>
		);
	}
}
MacroTracker.propTypes = {
	macros: {
		target: {
			protein: PropTypes.number,
			carbs: PropTypes.number,
			fat: PropTypes.number,
		},
		current: {
			protein: PropTypes.number,
			carbs: PropTypes.number,
			fat: PropTypes.number,
		},
	},
	setMacros: PropTypes.func.isRequired,
	updateMacro: PropTypes.func.isRequired,
	resetMacros: PropTypes.func.isRequired,
};
MacroTracker.defaultProps = {
	macros: {
		target: {
			protein: 0,
			carbs: 0,
			fat: 0,
		},
		current: {
			protein: 0,
			carbs: 0,
			fat: 0,
		},
	},
};

export default MacroTracker;
