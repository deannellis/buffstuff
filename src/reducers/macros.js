const macrosReducerDefaultState = {
	target: {
		protein: null,
		carbs: null,
		fat: null
	},
	current: {
		protein: null,
		carbs: null,
		fat: null
	},
	dateObject: null
};

const macrosReducer = (state = macrosReducerDefaultState, action) => {
	let setTargetUpdates = {};
	const { update } = action;
	const updateMacroUpdates = { ...state };
	const { currentDate } = action;
	const setCurrentDateUpdates = { ...state };
	const resetCurrentUpdates = {
		target: { ...state.target },
		current: { ...macrosReducerDefaultState.current },
		dateObject: null
	};
	switch (action.type) {
		case 'SET_TARGET_MACROS':
			setTargetUpdates = {
				target: { ...action.target },
				current: { ...state.current },
				dateObject: { ...state.dateObject }
			};
			return setTargetUpdates;
		case 'UPDATE_MACRO':
			if (updateMacroUpdates.current[update.macro] !== null) {
				updateMacroUpdates.current[update.macro] += update.amount;
			} else {
				updateMacroUpdates.current[update.macro] = update.amount;
			}
			return updateMacroUpdates;
		case 'SET_CURRENT_DATE':
			setCurrentDateUpdates.dateObject = currentDate;
			return setCurrentDateUpdates;
		case 'RESET_CURRENT':
			return resetCurrentUpdates;
		default:
			return state;
	}
};

export default macrosReducer;
