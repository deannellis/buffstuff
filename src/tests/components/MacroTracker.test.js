import React from 'react';
import { shallow } from 'enzyme';

import MacroTracker from '../../components/MacroTracker';
import { macros } from '../fixtures/macros';

test('should render Macro Tracker without props', () => {
    const wrapper = shallow(<MacroTracker />);
    expect(wrapper).toMatchSnapshot();
});

test('should render Macro Tracker with props', () => {
    const wrapper = shallow(<MacroTracker macros={macros} />);
    expect(wrapper).toMatchSnapshot();
});

// test('should open update form', () => {
//     const wrapper = shallow(<MacroTracker macros={macros} />);
//     wrapper.find('Button').at(0).simulate('click');
//     expect(wrapper.state('displayUpdateForm')).toBe(true);
// });