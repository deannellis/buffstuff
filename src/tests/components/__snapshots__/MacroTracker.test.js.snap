// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Macro Tracker with props 1`] = `
<div
  className="macro-tracker card"
>
  <div
    className="macro-tracker__header"
  >
    <h2>
      Macros
    </h2>
    <button
      className="macro-tracker__menu-button"
      onClick={[Function]}
      type="button"
    >
      <DotsIcon
        fill="#fff"
        size={24}
      />
    </button>
  </div>
  <PieChart
    data={
      Array [
        Object {
          "amount": 157,
          "macro": "protein",
        },
        Object {
          "amount": 130,
          "macro": "carbs",
        },
        Object {
          "amount": 80,
          "macro": "fat",
        },
        Object {
          "amount": 333,
          "macro": "leftover",
        },
      ]
    }
  />
  <div
    className="macro-tracker__macros"
  >
    <div
      className="macro-tracker__macro"
      key="protein"
    >
      <div
        className="macro-tracker__macro-key"
        id="macro-key-0"
      />
      <p>
        <span>
          protein
        </span>
         157 of 340 grams
      </p>
    </div>
    <div
      className="macro-tracker__macro"
      key="carbs"
    >
      <div
        className="macro-tracker__macro-key"
        id="macro-key-1"
      />
      <p>
        <span>
          carbs
        </span>
         130 of 280 grams
      </p>
    </div>
    <div
      className="macro-tracker__macro"
      key="fat"
    >
      <div
        className="macro-tracker__macro-key"
        id="macro-key-2"
      />
      <p>
        <span>
          fat
        </span>
         80 of 80 grams
      </p>
    </div>
  </div>
  <Button
    clickHandler={[Function]}
    disabled={false}
    type="button"
    variant="primary"
  >
    Add macros
  </Button>
  <Button
    clickHandler={[Function]}
    disabled={false}
    type="button"
    variant=""
  >
    reset macros
  </Button>
  <div
    className="macro-tracker__overlay card__overlay  card__overlay--hidden"
  >
    <UpdateMacroForm
      closeForm={[Function]}
      updateMacro={[Function]}
    />
  </div>
  <div
    className="macro-tracker__overlay card__overlay  card__overlay--hidden macro-tracker__menu"
  >
    <p>
      Update Target Macros
    </p>
    <SetMacrosForm
      current={
        Object {
          "carbs": 280,
          "fat": 80,
          "protein": 340,
        }
      }
      submitMacros={[Function]}
    />
    <div
      className="macro-tracker__menu-bottom"
    >
      <Button
        clickHandler={[Function]}
        disabled={false}
        type="button"
        variant=""
      >
        Close Menu
      </Button>
    </div>
  </div>
</div>
`;

exports[`should render Macro Tracker with required props 1`] = `
<div
  className="macro-tracker card"
>
  <div
    className="macro-tracker__header"
  >
    <h2>
      Macros
    </h2>
    <button
      className="macro-tracker__menu-button"
      onClick={[Function]}
      type="button"
    >
      <DotsIcon
        fill="#fff"
        size={24}
      />
    </button>
  </div>
  <PieChart
    data={
      Array [
        Object {
          "amount": 0,
          "macro": "protein",
        },
        Object {
          "amount": 0,
          "macro": "carbs",
        },
        Object {
          "amount": 0,
          "macro": "fat",
        },
      ]
    }
  />
  <div
    className="macro-tracker__macros"
  >
    <div
      className="macro-tracker__macro"
      key="protein"
    >
      <div
        className="macro-tracker__macro-key"
        id="macro-key-0"
      />
      <p>
        <span>
          protein
        </span>
         0 of 0 grams
      </p>
    </div>
    <div
      className="macro-tracker__macro"
      key="carbs"
    >
      <div
        className="macro-tracker__macro-key"
        id="macro-key-1"
      />
      <p>
        <span>
          carbs
        </span>
         0 of 0 grams
      </p>
    </div>
    <div
      className="macro-tracker__macro"
      key="fat"
    >
      <div
        className="macro-tracker__macro-key"
        id="macro-key-2"
      />
      <p>
        <span>
          fat
        </span>
         0 of 0 grams
      </p>
    </div>
  </div>
  <Button
    clickHandler={[Function]}
    disabled={false}
    type="button"
    variant="primary"
  >
    Add macros
  </Button>
  <Button
    clickHandler={[Function]}
    disabled={false}
    type="button"
    variant=""
  >
    reset macros
  </Button>
  <div
    className="macro-tracker__overlay card__overlay  card__overlay--hidden"
  >
    <UpdateMacroForm
      closeForm={[Function]}
      updateMacro={[Function]}
    />
  </div>
  <div
    className="macro-tracker__overlay card__overlay  card__overlay--hidden macro-tracker__menu"
  >
    <p>
      Update Target Macros
    </p>
    <SetMacrosForm
      current={
        Object {
          "carbs": 0,
          "fat": 0,
          "protein": 0,
        }
      }
      submitMacros={[Function]}
    />
    <div
      className="macro-tracker__menu-bottom"
    >
      <Button
        clickHandler={[Function]}
        disabled={false}
        type="button"
        variant=""
      >
        Close Menu
      </Button>
    </div>
  </div>
</div>
`;
